x/evidence/src/types/router.rs

```diff
@@ -0,0 +1,229 @@
+use crate::Evidence;
+use anyhow::anyhow;
+use gears::{
+    context::{block::BlockContext, init::InitContext, tx::TxContext},
+    store::{database::Database, StoreKey},
+};
+use std::{collections::HashMap, rc::Rc};
+
+#[derive(Clone)]
+#[allow(clippy::type_complexity)]
+pub struct HandlerFn<CTX, E>(Rc<dyn Fn(&mut CTX, &E) -> anyhow::Result<()>>);
+
+impl<CTX, E> HandlerFn<CTX, E> {
+    pub fn new<F: Fn(&mut CTX, &E) -> anyhow::Result<()> + 'static>(f: F) -> Self {
+        Self(Rc::new(f))
+    }
+
+    pub fn call(&self, ctx: &mut CTX, evidence: &E) -> anyhow::Result<()> {
+        self.0(ctx, evidence)
+    }
+}
+
+/// Handler defines an agnostic Evidence handler. The handler is responsible
+/// for executing all corresponding business logic necessary for verifying the
+/// evidence as valid. In addition, the Handler may execute any necessary
+/// slashing and potential jailing.
+// this ugly api is used because it allows to skip multiple ctx generics in declaration of keeper
+// and abci handler
+// we have to provide generics in the abci constructor and I think that it may be annoying
+//
+// TODO: think how to improve the router
+#[derive(Clone)]
+pub enum Handler<'a, DB, SK, E>
+where
+    E: Evidence,
+    SK: StoreKey,
+    DB: Database,
+{
+    Init(Rc<HandlerFn<InitContext<'a, DB, SK>, E>>),
+    Tx(Rc<HandlerFn<TxContext<'a, DB, SK>, E>>),
+    Block(Rc<HandlerFn<BlockContext<'a, DB, SK>, E>>),
+}
+
+impl<'a, DB, SK, E> Handler<'a, DB, SK, E>
+where
+    E: Evidence,
+    SK: StoreKey,
+    DB: Database,
+{
+    pub fn new_init(handler_fn: HandlerFn<InitContext<'a, DB, SK>, E>) -> Self {
```

Comment:
Hmmm. This all structure is... okay. What about implementing `From`?

Reply:
non actual